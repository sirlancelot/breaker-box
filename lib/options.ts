import type { CircuitBreakerOptions } from "./types.js"
import type { AnyFn } from "./util.js"

function assert(value: unknown, message?: string): asserts value {
	if (!value) throw new TypeError(message)
}

export function parseOptions<Fallback extends AnyFn>(
	options: CircuitBreakerOptions<Fallback>,
) {
	const {
		errorIsFailure = () => false,
		errorThreshold = 0,
		errorWindow = 10_000,
		minimumCandidates = 6,
		onClose,
		onOpen,
		resetAfter = 30_000,
		retryDelay = () => undefined,
	} = options

	// errorIsFailure
	assert(
		typeof errorIsFailure === "function",
		`"errorIsFailure" must be a function (received ${typeof errorIsFailure})`,
	)

	// errorThreshold
	assert(
		errorThreshold >= 0 && errorThreshold <= 1,
		`"errorThreshold" must be a number between 0 and 1 (received ${errorThreshold})`,
	)

	// errorWindow
	assert(
		errorWindow > 0,
		`"errorWindow" must be milliseconds greater than 0 (received ${errorWindow})`,
	)

	// minimumCandidates
	assert(
		minimumCandidates > 1,
		`"minimumCandidates" must be a number greater than 1 (received ${minimumCandidates})`,
	)

	// (optional) onClose
	if (onClose)
		assert(
			typeof onClose === "function",
			`"onClose" must be a function (received ${typeof onClose})`,
		)

	// (optional) onOpen
	if (onOpen)
		assert(
			typeof onOpen === "function",
			`"onOpen" must be a function (received ${typeof onOpen})`,
		)

	// resetAfter
	assert(
		resetAfter > 0,
		`"resetAfter" must be milliseconds greater than 0 (received ${resetAfter})`,
	)
	assert(
		resetAfter >= errorWindow,
		`"resetAfter" must be milliseconds greater than or equal to "errorWindow" (received ${resetAfter})`,
	)

	// retryDelay
	assert(
		typeof retryDelay === "function",
		`"retryDelay" must be a function (received ${typeof retryDelay})`,
	)

	return {
		errorIsFailure,
		errorThreshold,
		errorWindow,
		minimumCandidates,
		onClose,
		onOpen,
		resetAfter,
		retryDelay,
	}
}
