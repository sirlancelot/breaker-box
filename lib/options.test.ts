import { expect, it } from "vitest"
import { parseOptions } from "./options.js"

// Serialize functions by showing their source code
expect.addSnapshotSerializer({
	test: (value) => value instanceof Function,
	serialize: (value: Function) => JSON.stringify(value.toString()).slice(1, -1),
})

it("sets defaults", ({ expect }) => {
	const options = parseOptions({})
	expect(options).toMatchSnapshot()
})

it.for([
	{ prop: "errorIsFailure", value: 42 },
	{ prop: "errorThreshold", value: -1 },
	{ prop: "errorWindow", value: -1 },
	{ prop: "minimumCandidates", value: 0 },
	{ prop: "onClose", value: 42 },
	{ prop: "onOpen", value: 42 },
	{ prop: "resetAfter", value: -1 },
	{ prop: "retryDelay", value: 42 },
])("handles $prop error", ({ prop, value }, { expect }) => {
	expect(() => parseOptions({ [prop]: value })).toThrowErrorMatchingSnapshot()
})

it("handles resetAfter < errorWindow", ({ expect }) => {
	expect(() =>
		parseOptions({ resetAfter: 1_000, errorWindow: 2_000 }),
	).toThrowErrorMatchingSnapshot()
})
