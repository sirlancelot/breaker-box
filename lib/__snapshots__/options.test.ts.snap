// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles 'errorIsFailure' error 1`] = `[TypeError: "errorIsFailure" must be a function (received number)]`;

exports[`handles 'errorThreshold' error 1`] = `[TypeError: "errorThreshold" must be a number between 0 and 1 (received -1)]`;

exports[`handles 'errorWindow' error 1`] = `[TypeError: "errorWindow" must be milliseconds greater than 0 (received -1)]`;

exports[`handles 'minimumCandidates' error 1`] = `[TypeError: "minimumCandidates" must be a number greater than 1 (received 0)]`;

exports[`handles 'onClose' error 1`] = `[TypeError: "onClose" must be a function (received number)]`;

exports[`handles 'onOpen' error 1`] = `[TypeError: "onOpen" must be a function (received number)]`;

exports[`handles 'resetAfter' error 1`] = `[TypeError: "resetAfter" must be milliseconds greater than 0 (received -1)]`;

exports[`handles 'retryDelay' error 1`] = `[TypeError: "retryDelay" must be a function (received number)]`;

exports[`sets defaults 1`] = `
{
  "errorIsFailure": () => false,
  "errorThreshold": 0,
  "errorWindow": 10000,
  "minimumCandidates": 6,
  "onClose": undefined,
  "onOpen": undefined,
  "resetAfter": 30000,
  "retryDelay": () => void 0,
}
`;
